### PR TITLE
Find missing events

### LINKED ISSUE #319: Find lost events
In @klannon's recent [processing](http://www.crc.nd.edu/~klannon/tracklet_single_mu_withPU_v01/all/) run, Lobster thought he was done at 100k events even though DAS thinks the dataset has 168k. We need to understand this discrepancy-- are we skipping events? It's also possible that DAS is wrong. This dataset [1] is a good test case because it has many split lumis:

```
sqlite> select count(*) from (select distinct file, run, lumi from units_minus);
1681
sqlite> select count(*) from (select distinct run, lumi from units_minus);
1001
```

[1] /SingleMuMinusFlatPt0p2To150/TTI2023Upg14D-PU140bx25_PH2_1K_FB_V3-v2/GEN-SIM-DIGI-RAW

### PR DESCRIPTION
**Pull request by annawoodard**
We currently do '[lumi-mending](https://github.com/matz-e/lobster/blob/master/lobster/core/unit.py#L395)' of split lumis, where we pick a (run, lumi) that hasn't been finished, then do a second query to find all of the files with that (run, lumi) in it and add those to the processing of the current task. To try and debug the missing events issue, I wrote a little [script](https://github.com/matz-e/lobster/blob/find-missing-events/test/count_events.py) to count the unique events in a dataset.

Looking at the original dataset, to double check this isn't a problem in DAS (I copied it locally for faster debugging):

```
~/lobster/test >python count_events.py /hadoop/store/user/awoodard/minus
found 168000 (100000 unique) run, lumi, event sets in 192 files in path /hadoop/store/user/awoodard/minus
```

Here is the [result](http://www.crc.nd.edu/~awoodard/lobster/simple_crustacean_7be2ff1_v1/) of running with the master branch:

```
~/lobster/test >python count_events.py /hadoop/store/user/awoodard/simple_crustacean_7be2ff1_v1/minus/
found 100000 (100000 unique) run, lumi, event sets in 174 files in path /hadoop/store/user/awoodard/simple_crustacean_7be2ff1_v1/minus/
```

Here is the [result](http://www.crc.nd.edu/~awoodard/lobster/simple_crustacean_ba3a0a6_v1/) of running with the lumi-mending removed (this branch):

```
(.lobster)[earth] ~/lobster/test >python count_events.py /hadoop/store/user/awoodard/simple_crustacean_ba3a0a6_v1/minus/
found 167300 (100000 unique) run, lumi, event sets in 403 files in path /hadoop/store/user/awoodard/simple_crustacean_ba3a0a6_v1/minus/
```

So in @klannon's dataset, there are a large number of duplicate run, lumi, event sets. I don't think we want to throw events away just because they have duplicate run, lumi, event sets (for MC, at least.) I can think of innocuous scenarios leading to duplicates where the events themselves are perfectly useful (as long as the random seed is not fixed.) What is the use case we are trying to protect against-- mistakes upstream that result in real duplication of data? I think we need at least a knob to turn off lumi mending, although my vote would be to remove it completely (which this PR does) in order to keep things simpler. @matz-e, @klannon: opinions?

Fixes #319.

### CODE REVIEW

**Comment by annawoodard** [comment]
Oops, sorry about that. I'll fix the tests tomorrow.

**Comment by matz-e** [comment]
We had a very specific reason/corner-case of double-processing to put that check in:

* Assume that a dataset has split lumis
* We process two files in one task, each contains a certain `(run, lumi)`
* We take a unit with `(run, lumi)` from file one, but not from file two, because the task size is big enough
* Since CMSSW looks at the lumi mask without any file information, and processes both units
* The second unit will still be marked unprocessed in the database, since we did not intend for it to be processed

Really, I think these screwed up datasets may be better processed in file-based mode. In fact, I'd be happy to remove these checks as you have done here, but I think we should add a warning to the dataset class about split lumis and advise caution/refuse to use lumi-based splitting. I think with everything that has been done in 7xy, this should not occur anymore, but explicit checks would be better.
Alternatively, we can add a unit-based processing which stops on file-boundaries, and removes the corner-case (this is in WMCore as an option, not sure if the CRAB uses it).

That being said, I personally don't need that check, and hence really like this PR, because I'm done with everything released with `CMSSW < 7xy`, and I think the official chain is purely lumi-based by now. I think this will (a) simplify the code and (b) speed up job creation as we reduce the amount of queries. But I still think we should add a caveat warning…

Curious as to what @klannon's opinion is on this matter?

BTW, I think that the test look at exactly this corner-case from above and would have to be re-written better.

**TL;DR** (this normally goes at the bottom of posts, at least in the corners of the web where I hang out): Let's merge this and add a check for duplicate lumis to the `cmssw.Dataset` class. Then we should stop creating tasks at file-boundaries (taking only units from one file per task).

**Comment by annawoodard** [comment]
Thanks for explaining the corner case so clearly, I remember now. (Funny how the development goes in circles.)

> Let's merge this and add a check for duplicate lumis to the cmssw.Dataset class. Then we should stop creating tasks at file-boundaries (taking only units from one file per task).

Are you saying we should always stop at file boundaries (simpler, my preference) or only if the check for duplicate lumis has failed? If we always stop at file boundaries, do we need a check (is there some further danger the user needs to know about)?

**Comment by matz-e** [comment]
For the sake of completely adjustable tasks, I was thinking that stopping at file boundaries should only required when running with complex datasets. This was one of our main features (task size adjustability to match runtime specification), and I don't think we should give that up. Yes, it'll be a bit more complex (I think basically just vetoing to stop on encountering a new file), but I don't think any more complex than what we have currently. It would be easiest if we just required datasets with split lumis to be run on "file-based" units, but can potentially lead to very long processing times.

**Comment by annawoodard** [comment]
> For the sake of completely adjustable tasks, I was thinking that stopping at file boundaries should only required when running with complex datasets. 

By complex datasets, you mean, datasets with split lumis? So we do a check if `total_lumis` != `total_units` in `cmssw.Dataset`, then maybe set a flag in the workflow table that we can grab [here](https://github.com/matz-e/lobster/blob/master/lobster/core/unit.py#L301) telling us whether or not to limit the fileinfo query to 1 file [here](https://github.com/matz-e/lobster/blob/master/lobster/core/unit.py#L318)?

> This was one of our main features (task size adjustability to match runtime specification), and I don't think we should give that up.

Good point! I was being too aggressive in my quest for simplicity.

**Comment by annawoodard** [comment]
EDIT: That wasn't quite right because we wouldn't have to limit the query, but you get the idea.

**Comment by matz-e** [comment]
> By complex datasets, you mean, datasets with split lumis? So we do a check if `total_lumis` != `total_units` in `cmssw.Dataset`, then maybe set a flag in the workflow table to pick up [here](https://github.com/matz-e/lobster/blob/master/lobster/core/unit.py#L301) telling us whether or not to stop on file boundaries?

Sounds like a plan. Just looked through the code, didn't you add `total_lumis` somewhere already?

**Comment by annawoodard** [comment]
I traded `total_lumis` for `total_units`, but I can add it back in (we get it from the DBS API.) I'll go ahead and draft it up and add it to this branch for review. Do we have a preferred terminology for stopping on file boundaries (how do they call it in WMCore again?)

**Comment by matz-e** [comment]
It's [`halt_job_on_file_boundaries`](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/JobSplitting/EventAwareLumiBased.py#L45), which I think is a little bulky and not so clear. How about something like `single_file_mode` or `single_file_per_task`?

**Comment by klannon** [comment]
Sorry for the radio silence.  I don't think I have any great technical insights.  My general suggestions of favoring simplicity over being able to address every possible permutation automatically apply here.  I am a little torn because I suspect that in this case, Lobster is actually saving me from having duplicated events because I think that there are really only supposed to be 100k events and when you "fix" Lobster and the output goes to 168k, many of those will be duplicates.  Nonetheless, I don't think this is a problem that Lobster should be solving.  In *many* cases (as @annawoodard points out), different events are accidentally generated with identical run and lumi section numbers.  In those cases Lobster would prevent users from accessing the full stats of the sample.

So, I support implementing whatever is the simplest approach.  If that approach would lead to not processing a full dataset in some weird corner cases, then we should have a clear alternative option or setting for dealing with that.  Quick question about file-based processing: if my files live somewhere else (say via xrootd) can I specify the list of files in a text file?  I'm only seem to remember being able to do this by pointing Lobster at a local directory.

**Comment by matz-e** [comment]
@klannon, I think what @annawoodard brought up and what we're discussing here would lead to a simplification, without loosing the corner-cases. If your events are duplicated, you would have them in your output twice, though.

As you say, it's not really something Lobster should solve, and as long as you have EDM output, merging everything into one big file would get rid of duplicated `(run, lumi, event)` events.

Not sure about the file-based processing, there should be settings to use file-based processing with datasets from DBS. We should actually be able to do some xrootd query for datasets living elsewhere?

**Comment by annawoodard** [comment]
> As you say, it's not really something Lobster should solve, and as long as you have EDM output, merging everything into one big file would get rid of duplicated (run, lumi, event) events.

Ah, thanks for reminding me about this. Now I'm going in circles because, depending on your merging parameters, you could have a many-to-one mapping between input and output events unless we break merging. Maybe it's better if we just leave things the way they are (except add in a warning about split lumis and warn that if there are duplicated `(run, lumi, event)`s that the 'events' might not equal 'read' in the Lobster summary.) If events MAY be deduplicated it's probably better to always deduplicate (as we're doing now) than sometimes deduplicate in a hard-to-predict way, since never deduplicating isn't an option.

**Comment by klannon** [comment]
I guess the real question I have is, let's say right now I have some files out there somewhere (i.e. not local, but accessible via any of the usual mechanisms like xrootd, srm, or chirp) but they're not in DBS, is there a way to ask Lobster to process them file-by-file?  If not, then I'll make an issue requesting this mode.  If there is, then I think that plus the approach your discussing above handles every possible case, right?  And yes, I agree that if I've got duplication, that's my problem, not Lobster's.

Argh--too slow.  @annawoodard: I'm very confused about what you write.  I think we always want Lobster to assume it should process everything.  It should never try to detect and correct duplication in the input list.  Or am I missing something?

**Comment by annawoodard** [comment]
I guess the real question I have is, let's say right now I have some files out there somewhere (i.e. not local, but accessible via any of the usual mechanisms like xrootd, srm, or chirp) but they're not in DBS, is there a way to ask Lobster to process them file-by-file?

[This](http://lobster.readthedocs.org/en/latest/config.html#lobster.core.dataset.Dataset) should work, it takes lists or directories.

> I think we always want Lobster to assume it should process everything. It should never try to detect and correct duplication in the input list. Or am I missing something?

My point is that, if we don't correct duplication (which we're doing now), than some duplication will be corrected if your merge size is big enough and the two duplicated events end up in the same file. It would be better to always correct, so that the output is predictable, rather than have some events de-duplicated randomly.

**Comment by klannon** [comment]
Duplicate removal CMSSW is optional.  We should merge with the CMSSW 
parameter that doesn't remove duplicates.  We could allow duplicate 
removal as an advanced option.  Pure ROOT merging won't ever remove 
duplicates, so we should document the option clearly.

**Comment by matz-e** [comment]
@annawoodard, @klannon to use the `core.Dataset` class with XrootD, you may have to specify the input parameters of the `StorageConfiguration` right.

File: lobster/core/unit.py [hunk 1]
```diff
@@ -389,7 +392,7 @@ def insert_task(files, units, arg):
 
                 current_size += 1
 
-                if current_size == tasksize:
+                if (current_size == tasksize) or (single_file_max and len(files) > 1):
```
Comment (matz-e) [inline lobster/core/unit.py]: Not sure your logic is correct here. If `len(files) > 1` already, don't we have two files in the task, but we want to have one only?

Comment (annawoodard) [inline lobster/core/unit.py]: Thanks!

**Comment by annawoodard** [comment]
@matz-e: This isn't as important as the scale test preparations, so I'll return to it and fix the tests afterwards.

**Comment by annawoodard** [comment]
I do not feel as gung-ho as I did about this earlier. We have this in the history if we want to resurrect it later I don't think it's worth the danger of introducing new bugs right now.

**Comment by matz-e** [comment]
This clearly needs a rebase — I think it would be worth it. If we have datasets that stop at the file boundary, we can have datasets with "continuous" lumi requirements — enabling us to enable LHE tasks that can process several units at once, and give us **dynamic LHE task sizes**. Code changes are fairly minor for that, too. I could quickly hack up something and hand it over to @WenzhaoLi for verification, testing, and approval?

**Comment by annawoodard** [comment]
+1 sounds good to me

**Comment by klannon** [comment]
+1 from me too, but only if your advisor doesn't object to your spending time on this instead of your thesis.  I definitely like the idea of not having you do it all, but letting @WenzhaoLi get involved.

**Comment by WenzhaoLi** [comment]
Sorry for being late to the party. I will put this into my to-do-list.

**Comment by matz-e** [comment]
I can rebase this for you and add some required changes… that should be a good start.

**Comment by WenzhaoLi** [comment]
That'd be great! Thanks @matz-e !
